x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -494,25 +496,30 @@ private void markAsNeedsFSync() {
                         // Capture the completed ranges before fsyncing; ranges that are completed after this point won't be considered as
                         // persisted on disk by the caller of this method, even if they are fully written to disk at the time the file
                         // fsync is effectively executed
-                        final SortedSet<Tuple<Long, Long>> completedRanges = tracker.getCompletedRanges();
-                        assert completedRanges != null;
-                        assert completedRanges.isEmpty() == false;
+                        final SortedSet<Tuple<Long, Long>> fsyncedRanges = tracker.getCompletedRanges();
+                        assert fsyncedRanges != null;
+                        assert fsyncedRanges.isEmpty() == false;
 
-                        IOUtils.fsync(file, false, false);
+                        if (evicted.get() == false) {
+                            IOUtils.fsync(file, false, false);
+                            completedRanges = fsyncedRanges;
+                        }
                         success = true;
-                        return completedRanges;
                     } finally {
                         if (success == false) {
                             markAsNeedsFSync();
                         }
                     }
                 }
             } finally {
-                decrementRefCount();
+                if (decrementRefCount()) {
+                    // cache file got evicted during fsync (maybe a shard eviction?)
+                    completedRanges = Collections.emptySortedSet();
```

Comment:
I am not sure this solves the issue, since the evict could happen right after this and thus result in the thread running here reinserting the data?

I wonder if we could track the deletions on the writer until commit and ignore any updates targetting deleted data there? Since we only commit from the single cache-sync thread, that might work? It could be a bit sloppy in that commit should simply ensure that the deletes that was in the map before committing are removed - only for memory cleanup reasons. Whether a delete registration survives a commit round inadvertently is not important, freeing us from synchronizing the entire commit against the delete.

Reply:
> I am not sure this solves the issue, since the evict could happen right after this and thus result in the thread running here reinserting the data?

I'm afraid you're right; a slightly modification in the test I added can catch the case you mentioned.

> I wonder if we could track the deletions on the writer until commit and ignore any updates targetting deleted data there? Since we only commit from the single cache-sync thread, that might work?

I also thought about this but abandoned it due to the extra complexity it adds. I'm going to investigate this solution again.

Reply:
@henningandersen Thanks again for your feedback on this.

> I wonder if we could track the deletions on the writer until commit and ignore any updates targetting deleted data there? Since we only commit from the single cache-sync thread, that might work? It could be a bit sloppy in that commit should simply ensure that the deletes that was in the map before committing are removed - only for memory cleanup reasons. Whether a delete registration survives a commit round inadvertently is not important, freeing us from synchronizing the entire commit against the delete.

I tried to implement this idea but didn't manage to come through it without adding some kind of synchronization between deletes and updates at the writer level. I think that the way cache files work today does not guarantee that no cache file can be updated after eviction started and before the file is fully released and deleted. It is even more likely when writing a cache file range is throttled due to indices recovery rate limit.

In 6199a23 I've update the test I made before to make it fail in the scenario you described. In 6199a23 I've updated the code to introduce another way to fix the initial issue of cache synchronization reindexing evicted cache files back in the persistent cache index.

The solution I propose is to ensure that marking a cache file as "need fsync" is always done before the cache file reference is released; and then when the cache file is fully released it is marked as "deleted". Those events are now tracked in a queue of cache files events (the previous `cacheFilesToSync` becomes `ConcurrentLinkedQueue<CacheFileEvent> cacheFilesEventsQueue`) whose FIFO nature should prevent an "update/need fsync" event to come after a "delete" event. This list of events is then processed sequentially by the cache synchronization method.

I think this solution  prevents the cache synchronization to add evicted cache information back in the persistent cache. In my opinion it also simplifies how cache files updates/deletes and persistent cache are interacting together. It also centralizes the persistent cache indexing at a single place in the `synchronizeCache()` method. 

The only inconvenience I can think of is the persistent cache get cache size method that can be out of sync (ie reports a non-zero size for evicted shards) until the next cache synchronization is executed. I think we can address this as a follow-up (using a delete by query against the persistent cache to delete all docs related to the evicted shard).

@henningandersen I'd be very happy if you can give this another look and tell me if you find a flaw in the solution I propose. :pray:

Reply:
Thanks @tlrx , I do like this approach. I think the cache size issue can also simply be fixed by checking that the file exists when calculating the cache size.

Is this ready for another round of review? I only looked at the main bits of the change you highlighted above.

Reply:
> I think the cache size issue can also simply be fixed by checking that the file exists when calculating the cache size.

That sounds like a good idea, I gave it a try in 626d70c